File: my-solutions/0238-product-of-array-except-self/solution.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        res = [1] * len(nums)
        left = [1] * (len(nums) + 1)
        for i in range(1, len(left)):
            left[i] = nums[i-1] * left[i-1]
        # print(left)

        right = [1] * (len(nums) + 1)
        for i in range(len(right) - 2, -1, -1):
            right[i] = nums[i] * right[i+1]
        # print(right)

        for i in range(len(res)):
            res[i] = left[i] * right[i+1]

        return res

File: my-solutions/0238-product-of-array-except-self/solution.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
       vector<int> lp(size(nums));
       vector<int> rp(size(nums));
       vector<int> answer(size(nums));

       lp[0] = nums[0];
       rp[size(nums) - 1] = nums[size(nums) - 1];

       for(int i = 1; i < size(nums); i++) {
           int j = size(nums) - 1 - i;
           lp[i] = lp[i-1] * nums[i];
           rp[j] = rp[j+1] * nums[j];
       }

       for(int i = 0; i < size(nums); i++) {
           if(i == 0) {
               answer[i] = rp[i+1];
           }  else if(i == size(nums) - 1) {
               answer[i] = lp[i-1];
           } else {
               answer[i] = lp[i-1] * rp[i+1];
           }
       }

       return answer;
    }
};
